warehouse/templates/includes/file-details.html

```diff
@@ -0,0 +1,134 @@
+    <li>{% trans tags=', '.join(file.pretty_wheel_tags) %} Tags: {{ tags }} {% endtrans %}</li>
+    {% if file.uploaded_via %}
+    <li>{% trans uploaded_via=file.uploaded_via %} Uploaded via: {{ uploaded_via }} {% endtrans %}</li>
+    {% endif %}
+  </ul>
+</div>
+
+<h3>File hashes</h3>
+<div>
+  <table class="table table--hashes">
+    <caption class="sr-only">{% trans filename=file.filename %}Hashes for {{ filename }}{% endtrans %}</caption>
+    <thead>
+      <tr>
+        <th scope="col">{% trans %}Algorithm{% endtrans %}</th>
+        <th scope="col">{% trans %}Hash digest{% endtrans %}</th>
+        <th></th>
+      </tr>
+    </thead>
+    <tbody>
+      <tr data-controller="clipboard">
+        <th scope="row">SHA256</th>
+        <td><code data-clipboard-target="source">{{ file.sha256_digest }}</code></td>
+        <td class="table__align-right">
+          <button type="button" class="button button--small copy-tooltip copy-tooltip-w" data-action="clipboard#copy" data-clipboard-target="tooltip" data-clipboard-tooltip-value="{% trans %}Copy to clipboard{% endtrans %}">
+            {% trans %}Copy{% endtrans %}
+          </button>
+        </td>
+      </tr>
+      <tr data-controller="clipboard">
+        <th scope="row">MD5</th>
+        <td><code data-clipboard-target="source">{{ file.md5_digest }}</code></td>
+        <td class="table__align-right">
+          <button type="button" class="button button--small copy-tooltip copy-tooltip-w" data-action="clipboard#copy" data-clipboard-target="tooltip" data-clipboard-tooltip-value="{% trans %}Copy to clipboard{% endtrans %}">
+            {% trans %}Copy{% endtrans %}
+          </button>
+        </td>
+      </tr>
+      <tr data-controller="clipboard">
+        <th scope="row">BLAKE2b-256</th>
+        <td><code data-clipboard-target="source">{{ file.blake2_256_digest }}</code></td>
+        <td class="table__align-right">
+          <button type="button" class="button button--small copy-tooltip copy-tooltip-w" data-action="clipboard#copy" data-clipboard-target="tooltip" data-clipboard-tooltip-value="{% trans %}Copy to clipboard{% endtrans %}">
+            {% trans %}Copy{% endtrans %}
+          </button>
+        </td>
+      </tr>
+    </tbody>
+  </table>
+  <p>
+    {% trans href='https://pip.pypa.io/en/stable/topics/secure-installs/#hash-checking-mode', title=gettext('External link') %}
+      <a href="{{ href }}" title="{{ title }}" target="_blank" rel="noopener">See more details on using hashes here.</a>
+    {% endtrans %}
+  </p>
+</div>
+
+<h3>Attestations</h3>
+<div>
+  <p>
+    {% trans is_tp="Yes" if file.uploaded_via_trusted_publisher else "No" %}
+    Uploaded using Trusted Publishing? {{ is_tp }}
+    {% endtrans %}
+  </p>
+</div>
+
+{% if provenance %}
+<div>
+  <p>
+    The following attestations were made for <code>{{ file.filename }}</code>:
+  </p>
+
+  {% for bundle in provenance.attestation_bundles %}
```

Comment:
FWIW this doesn't quite work: `provenance` is now unbound, and `file.provenance` is now the raw JSON column rather than a `Provenance` model. I'll look at adjusting for this.

Reply:
Done with 71a4036561fe9c9e965643f05f2672c26c9cb14d: I've added `Provenance.as_model` to return the structured `pypi_attestations.Provenance` representation. I'm not sure if this is the best approach, though.